alf/networks/encoding_networks.py

```diff
@@ -230,6 +232,162 @@ def __init__(self,
+        Auto-shape inference: instead of specifying an initial start shape for
+        image deconv, this class only needs to specify the desired output shape
+        for the image and will automatically calculate the desired shape to start
+        decoding based on the specified ``transconv_layer_params``
+        and uses a FC layer to map the to the desired start shape.
+
+        Args:
+            input_size (int): the size of the input latent vector
+            transconv_layer_params (tuple[tuple]): a non-empty
+                tuple of tuple (num_filters, kernel_size, strides, padding),
+                where ``padding`` is optional.
+            output_shape (tuple): the complete output size would be
+                output_shape = (c, h, w).
+            start_decoding_channels (int): the initial number of channels we'd
+                like to have for the feature map. Note that we always first
+                project an input latent vector into a vector of an appropriate
+                length so that it can be reshaped into (``start_decoding_channels``,
+                ``start_decoding_height``, ``start_decoding_width``),
+                where ``start_decoding_height`` and ``start_decoding_width``
+                are automatically inferred based on the specified ``output_shape``
+                and ``transconv_layer_params``.
+            preprocess_fc_layer_params (tuple[int]): a tuple of fc
+                layer units. These fc layers are used for preprocessing the
+                latent vector before transposed convolutions.
+            activation (nn.functional): activation for hidden layers
+            kernel_initializer (Callable): initializer for all the layers.
+            output_activation (nn.functional): activation for the output layer.
+                Usually our image inputs are normalized to [0, 1] or [-1, 1],
+                so this function should be ``torch.sigmoid`` or
+                ``torch.tanh``.
+            name (str):
+        """
+        assert len(output_shape) == 3, "the output_shape should be (c, h, w)"
+        assert output_shape[0] == transconv_layer_params[-1][0], (
+            "channel number mis-match")
+
+        # compute conv shape and padding shape
+        conv_shapes = []
+        out_paddings = []
+        out_shape = output_shape[1:]
+        for i, paras in enumerate(transconv_layer_params[::-1]):
+            filters, kernel_size, stride = paras[:3]
+            kernel_size = common.tuplify2d(kernel_size)
+
+            padding = paras[3] if len(paras) > 3 else 0
+            padding = common.tuplify2d(padding)
+            conv_shape = self._calc_conv_out_shape(out_shape, padding,
+                                                   kernel_size, stride)
+            out_padding = self._calc_output_padding_shape(
+                out_shape, conv_shape, padding, kernel_size, stride)
+            out_shape = conv_shape
+            conv_shapes.append(conv_shape)
```

Comment:
No need to keep all conv shapes because only the last one will be used?

Reply:
Good point. Removed.